Dedup src diffs between 2.10 & 2.11+

dwijnand opened the pull request:
Fixes #233

dwijnand committed:
Dedup src diffs between 2.10 & 2.11+

Fixes #233

dwijnand commented:
I dropped https://github.com/sbt/zinc/pull/244/commits/a947ac8cd9bb891f8df05a23eae8c000fed6f054 which makes reviewing 2.10 Compat changes much easier.

jvican commented on internal/compiler-bridge/src/main/scala_2.10/xsbt/Compat.scala:
```diff
@@ -76,6 +79,20 @@ abstract class Compat {
+      sym.setter(base, hasExpandedName)
+
+    // copied from 2.12.1 sources
+    def needsExpandedSetterName: Boolean = (
```
Can we make this private?

dwijnand replied:
Sure I can. I made it public because it's public in the original (2.11+) API.

jvican commented on internal/compiler-bridge/src/main/scala_2.10/xsbt/Compat.scala:
```diff
@@ -76,6 +79,20 @@ abstract class Compat {
+    )
+
+    // unexpandedName replaces originalName in 2.11
+    def unexpandedName: Name = sym.originalName
```
`@inline` here perhaps? Scalac is very pessimistic about inlining, but it's better to have it than not to have it.

jvican commented on internal/compiler-bridge/src/main/scala_2.10/xsbt/Compat.scala:
```diff
@@ -76,6 +79,20 @@ abstract class Compat {
     def asMethod: MethodSymbol = sym.asInstanceOf[MethodSymbol]
+
+    // Not present in 2.10
+    final def getterIn(base: Symbol): Symbol = sym.getter(base)
```
Can we annotate these two also with `@inline`? Just in case, same comment as `unexpandedName`.

jvican commented on internal/compiler-bridge/src/main/scala_2.10/xsbt/Compat.scala:
```diff
@@ -101,41 +124,33 @@ abstract class Compat {
 }
+
+object Compat {
+  implicit final class TreeOps(val tree: sri.Trees#Tree) extends AnyVal {
```
Good work here.

jvican reviewed (approved):
This is great stuff, thanks for it @dwijnand.
  
Changes are much less intrusive of what I thought, and if they happen they will only affect 2.10 sources, so this patch has my blessing. I don't think we need benchmarks to get this in.
  
I'll rebase my other PR on top of this one, I really don't want to migrate all my stuff to 2.10.

dwijnand reviewed (commented):
I'll make the forwarders final and inline.

dwijnand commented:
This is green now, after a few brain farts and a moment of IO flakiness.

eed3si9n commented:
Nice work. Could you rebase against the master plz?

romanowski commented on internal/compiler-bridge/src/main/scala_2.10/xsbt/Compat.scala:
```diff
@@ -101,41 +124,33 @@ abstract class Compat {
-        }
-      }
-    }
+  implicit class MacroExpansionAttachmentCompat(self: MacroExpansionAttachment) {
```
extends AnyVal?

dwijnand replied:
Value classes can't be nested, and MacroExpansion depends on global.

romanowski replied:
ah, ok

romanowski reviewed (approved):
Good work!

dwijnand commented:
@jvican ExtractUsedNamesSpecification is failing. Do you mind having a look?

jvican commented:
Not at all! @dwijnand Will keep you posted tomorrow, I won't be very available but I'll try to have a look at this.

jvican commented:
This is weird. I cannot reproduce locally from your branch.

jvican commented:
No, wait, just reproduced.

jvican committed:
Fix issue with 2.10 traversals

This issue was fixed in https://github.com/sbt/zinc/pull/239/files#diff-65c95a9d18dc8a76c6182e1c6377fc65R154.

For some reason, 2.10 compiler is detecting root packages to have
symbols fields (which 2.11 do not recognise as such), so we need to
protect from them in `ExtractUsedNames` since `_root_` should definitely
not be registered as a name, because it's present in any Scala source
file.

jvican commented:
Just opened https://github.com/dwijnand/sbt-zinc/pull/2 to merge my fixes into dwijand to pass the CI here.

jvican commented:
CI passing. Waiting for another Zinc maintainer to sign off.
